### DISCUSSION TITLE
Using pnpm as a package manager when building in a monorepo

### DISCUSSION THREAD

**Discussion by V-ed** [dokku/dokku#5124] [Q&A] [answered]
This question is probably more around how Buildpacks work, and more than probably stems from my lack of knowledge on how Heroku works.

Essentially, I want to use PNPM as the package manager to install dependencies and potentially run the start script too (I am using a Procfile, so running the start command using pnpm would be easy to change), but I am not sure how I would go on about to install pnpm before the buildpack installs the dependencies and builds the application.

I tried to look into using different buildpacks, such as using this one (which seems kinda new actually) : https://github.com/unfold/heroku-buildpack-pnpm , but to no avail - a deploy seems to say there are multiple buildpacks and that the first one will be used, that it downloads the buildpack, but doesn't seem to do anything about it as the rest of the logs are unchanged :
```
       -----> Warning: Multiple default buildpacks reported the ability to handle this app. The first buildpack in the list below will be used.
              Detected buildpacks: multi nodejs
       -----> Multipack app detected
=====> Downloading Buildpack: https://github.com/unfold/heroku-buildpack-pnpm
```

and even though it seems to have downloaded the buildpack, looking at the buildpack's code, it still installs pnpm, simply doesn't install with pnpm if no lockfile is found (see point below), but pnpm should still be installed - which is isn't.

---

One seemingly important detail is that I am in a monorepo setup, therefore I am using the BUILD_DIR using the command `dokku builder:set $APP build-dir`, and I only have a root-level pnpm-lock, which I assume it means that the buildpacks won't see a pnpm-lock file in the building directory, which prevents logic in buildpacks to detect if pnpm should be used or not.

Is a solution to this is to simply configure pnpm to create a lockfile for each package (there's a setting for that)? Most docs says that creating multiple lockfiles isn't ideal as it degrades performances...

---

Basically, my question boils down to me not understanding how buildpacks work and maybe some pointers on how to use custom tools (pnpm) / do custom tasks (manually installing pnpm if possible) by using buildpacks...

**Comment by josegonzalez** [comment, marked as answer]
Folders and files in a monorepo setup get moved around early - so files must be relative to any specified build-dir (or app root if there isn't any). You'll need to have the pnpm files in that build-dir if you want the custom buildpack to work correctly.

**Reply by V-ed** [reply]
Perfect, that answers one of my questions and clears up any confusion about the build dir option, thanks for that!

I'm still unsure how I would go on about using pnpm in a monorepo however...

I've looked at the docs and I think I might have success into using a Dockerfile as a buildpack when paired with a Procfile.
Does that sound like a possible solution to using pnpm as a build tool, and if so, does the `release` step still run before running the `web` step of the Procfile file?

**Reply by josegonzalez** [reply]
You could move your pnpm files into the subdirectory you are attempting to build from _or_ stop trying to change the build dir (not sure what you are gaining by doing that if pnpm is supposed to be handling monorepo building).

Yes, the `release` process of a Procfile is still run for Dockerfile deploys.

**Reply by V-ed** [reply]
I think this clears up most of the remaining questions and I think that my best bet to keep my current setup and prevent having big apps (I don't want Dokku to contain the files from my frontend, I only use it for my backend, and not using the build dir would make the apps much bigger) is to use a Dockerfile.

Do you have any example repo that uses a Dockerfile for deploys? I never really used Dockerfiles so examples would be very appreciated, otherwise I'll simply do more research!

**Reply by V-ed** [reply]
Having an example to follow would be especially useful to see how cache is handled (to cache the pnpm store, for a concrete example).

I've already marked this as the answer but having pointers is always useful :)

**Reply by josegonzalez** [reply]
I think Dockerfile examples targeted to specific tooling is a bit out of the scope for the Dokku project. There is plenty of documentation online on constructing Dockerfiles for specific tools (and maybe one from the pnpm community, as they'd know best?). Seems like there is a long-ish issue here re: [pnpm and dockerfiles](https://github.com/pnpm/pnpm/issues/3114).

That said, if you'd like me to work on example code, consider [sponsoring my time](https://github.com/dokku/.github/blob/master/SPONSORING.md) for this particular project.

**Reply by V-ed** [reply]
Indeed, I didn't hope to get an already made example, but just pointers especially because I'm not sure how I'll handle caching in a Dockerfile in the context of Dokku's file.

With that said, I'll do some tests and share back here what I'll have found! Thanks for the pointer on that pnpm/dockerfile issue, I'll make sure to read as much as I can haha

Although I make no money from my projects, I'll also consider sponsoring Dokku, as it does help a lot with automating self-hosting solutions, and I do like it a lot :)
